File: sdk/flutter/praxy_core/lib/src/query.dart
import 'dart:convert';

/// A typed column reference. Carries no behaviour beyond its wire [name] — [T] only
/// exists so [Query.equal] and friends reject the wrong value type at compile time.
final class Col<T> {
  const Col(this.name);

  final String name;

  @override
  String toString() => name;
}

/// One query, kept as a value object end to end rather than pre-JSON-encoded —
/// unlike Appwrite's `Query.equal()` (which returns a `String`, forcing `Query.or()`
/// to `jsonDecode` its own inputs), a Praxy [Query] can be inspected, composed, and
/// encoded exactly once, at the point it's sent. Matches every method in
/// architecture.md §4.6 / research/appwrite-api.md's JSON-per-query wire format:
/// `{"method":"equal","attribute":"title","values":["Hello"]}`.
final class Query {
  const Query._(this.method, {this.attribute, this.values = const [], this.children = const []});

  final String method;
  final String? attribute;
  final List<Object?> values;
  final List<Query> children;

  static Query equal<T>(Col<T> col, T value) =>
      Query._('equal', attribute: col.name, values: [_encode(value)]);

  /// `equal` with multiple values is an OR within that one attribute (matches the
  /// server's `count >= 1` arity — a single [equal] call is just the `count == 1` case).
  static Query equalAny<T>(Col<T> col, List<T> values) =>
      Query._('equal', attribute: col.name, values: values.map(_encode).toList());

  static Query notEqual<T>(Col<T> col, T value) =>
      Query._('notEqual', attribute: col.name, values: [_encode(value)]);

  static Query lessThan<T>(Col<T> col, T value) =>
      Query._('lessThan', attribute: col.name, values: [_encode(value)]);

  static Query lessThanEqual<T>(Col<T> col, T value) =>
      Query._('lessThanEqual', attribute: col.name, values: [_encode(value)]);

  static Query greaterThan<T>(Col<T> col, T value) =>
      Query._('greaterThan', attribute: col.name, values: [_encode(value)]);

  static Query greaterThanEqual<T>(Col<T> col, T value) =>
      Query._('greaterThanEqual', attribute: col.name, values: [_encode(value)]);

  static Query between<T>(Col<T> col, T start, T end) =>
      Query._('between', attribute: col.name, values: [_encode(start), _encode(end)]);

  static Query isNull(Col<dynamic> col) => Query._('isNull', attribute: col.name);

  static Query isNotNull(Col<dynamic> col) => Query._('isNotNull', attribute: col.name);

  static Query startsWith(Col<String> col, String value) =>
      Query._('startsWith', attribute: col.name, values: [value]);

  static Query endsWith(Col<String> col, String value) =>
      Query._('endsWith', attribute: col.name, values: [value]);

  static Query contains<T>(Col<T> col, T value) =>
      Query._('contains', attribute: col.name, values: [_encode(value)]);

  /// Requires a declared fulltext index on [col] — the server rejects `search`
  /// without one rather than silently falling back to `ILIKE`.
  static Query search(Col<String> col, String value) =>
      Query._('search', attribute: col.name, values: [value]);

  static Query select(List<Col<dynamic>> columns) =>
      Query._('select', values: columns.map((c) => c.name).toList());

  static Query orderAsc(Col<dynamic> col) => Query._('orderAsc', attribute: col.name);

  static Query orderDesc(Col<dynamic> col) => Query._('orderDesc', attribute: col.name);

  static Query limit(int count) => Query._('limit', values: [count]);

  static Query offset(int count) => Query._('offset', values: [count]);

  static Query cursorAfter(String rowId) => Query._('cursorAfter', values: [rowId]);

  static Query cursorBefore(String rowId) => Query._('cursorBefore', values: [rowId]);

  static Query and(List<Query> queries) => Query._('and', children: queries);

  static Query or(List<Query> queries) => Query._('or', children: queries);

  /// The escape hatch for a method this builder doesn't cover yet — stays a real
  /// [Query] value like every other constructor, never a raw string.
  static Query raw(String method, {String? attribute, List<Object?> values = const []}) =>
      Query._(method, attribute: attribute, values: values);

  Map<String, dynamic> toJson() => {
    'method': method,
    if (attribute != null) 'attribute': attribute,
    if (children.isNotEmpty)
      'values': children.map((c) => c.toJson()).toList()
    else if (values.isNotEmpty)
      'values': values,
  };

  /// The wire form sent as one entry of the repeated `queries[]` param.
  String encode() => jsonEncode(toJson());

  static Object? _encode(Object? value) => switch (value) {
    DateTime d => d.toUtc().toIso8601String(),
    _ => value,
  };
}

File: sdk/js/packages/core/src/query.ts
/**
 * `Query`/`Col<T>` are value objects, not pre-encoded strings — mirrors `praxy_core`'s
 * `query.dart` (see `docs/research/nextjs-sdk.md`'s landmines: `Query.equal()` returning a raw
 * string forces `Query.or()` to re-decode its own inputs; composing child `Query` values directly
 * avoids that). Each `Query` becomes one entry of a repeated `queries[]` query-string param.
 */

/** A typed reference to a table column, used only to carry the column name at the type level. */
export class Col<T> {
  constructor(readonly name: string) {}
  // Phantom usage so `T` isn't reported as unused by the compiler.
  declare private readonly __type: T;
}

export type QueryJson = {
  method: string;
  attribute?: string;
  values?: unknown[];
};

export class Query {
  private constructor(
    readonly method: string,
    readonly attribute?: string,
    readonly values: unknown[] = [],
    readonly children: Query[] = [],
  ) {}

  static equal<T>(col: Col<T>, value: T): Query {
    return new Query("equal", col.name, [encode(value)]);
  }

  static equalAny<T>(col: Col<T>, values: T[]): Query {
    return new Query("equal", col.name, values.map(encode));
  }

  static notEqual<T>(col: Col<T>, value: T): Query {
    return new Query("notEqual", col.name, [encode(value)]);
  }

  static lessThan<T>(col: Col<T>, value: T): Query {
    return new Query("lessThan", col.name, [encode(value)]);
  }

  static lessThanEqual<T>(col: Col<T>, value: T): Query {
    return new Query("lessThanEqual", col.name, [encode(value)]);
  }

  static greaterThan<T>(col: Col<T>, value: T): Query {
    return new Query("greaterThan", col.name, [encode(value)]);
  }

  static greaterThanEqual<T>(col: Col<T>, value: T): Query {
    return new Query("greaterThanEqual", col.name, [encode(value)]);
  }

  static between<T>(col: Col<T>, start: T, end: T): Query {
    return new Query("between", col.name, [encode(start), encode(end)]);
  }

  static isNull(col: Col<unknown>): Query {
    return new Query("isNull", col.name);
  }

  static isNotNull(col: Col<unknown>): Query {
    return new Query("isNotNull", col.name);
  }

  static startsWith(col: Col<string>, value: string): Query {
    return new Query("startsWith", col.name, [value]);
  }

  static endsWith(col: Col<string>, value: string): Query {
    return new Query("endsWith", col.name, [value]);
  }

  static contains<T>(col: Col<T>, value: T): Query {
    return new Query("contains", col.name, [encode(value)]);
  }

  /** Requires a fulltext index on `col` server-side. */
  static search(col: Col<string>, value: string): Query {
    return new Query("search", col.name, [value]);
  }

  static select(columns: Col<unknown>[]): Query {
    return new Query("select", undefined, columns.map((c) => c.name));
  }

  static orderAsc(col: Col<unknown>): Query {
    return new Query("orderAsc", col.name);
  }

  static orderDesc(col: Col<unknown>): Query {
    return new Query("orderDesc", col.name);
  }

  static limit(count: number): Query {
    return new Query("limit", undefined, [count]);
  }

  static offset(count: number): Query {
    return new Query("offset", undefined, [count]);
  }

  static cursorAfter(rowId: string): Query {
    return new Query("cursorAfter", undefined, [rowId]);
  }

  static cursorBefore(rowId: string): Query {
    return new Query("cursorBefore", undefined, [rowId]);
  }

  static and(children: Query[]): Query {
    return new Query("and", undefined, [], children);
  }

  static or(children: Query[]): Query {
    return new Query("or", undefined, [], children);
  }

  /** Escape hatch for a method this builder doesn't wrap yet — still a real value object. */
  static raw(method: string, options: { attribute?: string; values?: unknown[] } = {}): Query {
    return new Query(method, options.attribute, options.values ?? []);
  }

  toJSON(): QueryJson {
    const json: QueryJson = { method: this.method };
    if (this.attribute !== undefined) json.attribute = this.attribute;
    if (this.children.length > 0) {
      json.values = this.children.map((c) => c.toJSON());
    } else if (this.values.length > 0) {
      json.values = this.values;
    }
    return json;
  }

  /** One entry of the repeated `queries[]` query-string param. */
  encode(): string {
    return JSON.stringify(this.toJSON());
  }
}

function encode(value: unknown): unknown {
  return value instanceof Date ? value.toISOString() : value;
}
